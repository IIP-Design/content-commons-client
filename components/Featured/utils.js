import { capitalizeFirst } from 'lib/utils';
import { normalizeItem } from 'lib/elastic/parser';
import { postTypeAggRequest } from 'lib/elastic/api';

/**
 * Get a string of up to three categories from a content item
 * @param {Object} item Normalized data for a give content item
 * @returns {string} Lowercase list of category names separated by ·
 */
export const getCategories = item => {
  const cats = item?.categories?.slice( 0, 3 ).reduce( ( acc, cat, index, arr ) => {
    const c = acc + cat.name.toLowerCase();

    return index < arr.length - 1 && index < 2 ? `${c} · ` : c;
  }, '' );

  return cats;
};

/**
 * Updates the the featured context
 * @param {Promise[]} array Group of promises that resolve the recents properties
 * @param {dispatch} dispatch A reducer dispatch function
 */
export const getFeatured = async ( array, dispatch ) => {
  const resArr = await Promise.all( array ).catch( err => {
    dispatch( {
      type: 'LOAD_FEATURED_FAILED',
    } );
  } );

  console.dir( resArr );

  const priorities = {};
  const recents = {};
  let items;

  if ( resArr ) {
    resArr.forEach( res => {
      if ( res?.data?.hits?.hits ) {
        items = res.data.hits.hits.map( item => normalizeItem( item, res.locale ) );

        switch ( res.component ) {
          case 'priorities':
            priorities[res.term] = items;
            break;
          case 'packages':
          case 'recents':
            recents[res.postType] = items;
            break;
          default:
            break;
        }
      }
    } );

    dispatch( {
      type: 'LOAD_FEATURED_SUCCESS',
      payload: {
        priorities,
        recents,
      },
    } );
  }
};


/**
 *
 * @param {dispatch} dispatch A reducer dispatch function
 * @param {Object} user User date object
 */
export const loadPostTypes = async ( dispatch, user ) => {
  dispatch( { type: 'LOAD_POST_TYPES_PENDING' } );

  let response;

  try {
    response = await postTypeAggRequest( user );
  } catch ( err ) {
    dispatch( { type: 'LOAD_POST_TYPES_FAILED' } );

    return;
  }

  if ( response?.aggregations?.postType?.buckets ) {
    const { buckets } = response.aggregations.postType;

    const filtered = buckets.filter( type => type.key !== 'courses' && type.key !== 'page' && type.key !== 'package' );

    const payload = filtered.map( type => {
      let displayName = capitalizeFirst( type.key );

      if ( type.key === 'post' ) displayName = 'Article';
      if ( type.key === 'document' ) displayName = 'Press Releases and Guidance';

      return {
        key: type.key,
        display_name: displayName,
        count: type.doc_count,
      };
    } );

    dispatch( {
      type: 'LOAD_POST_TYPES_SUCCESS',
      payload,
    } );
  } else {
    dispatch( {
      type: 'LOAD_POST_TYPES_SUCCESS',
      payload: [],
    } );
  }
};
