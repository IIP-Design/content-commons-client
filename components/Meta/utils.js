/**
 * Check if content exists for a given meta property
 * If not, filters it from the meta object
 *
 * @param {Object} meta All possible meta tags
 * @returns {Object} Only those meta tags for which there is content
 */
const filterEmpty = meta => {
  const filteredMeta = {};
  const metaArr = Object.entries( meta );

  metaArr.forEach( item => {
    if ( item[1] ) {
      const property = item[0];
      const value = item[1];

      filteredMeta[property] = value;
    }
  } );

  return filteredMeta;
};

/**
 * Converts the content item object into a metavalues array using the checkAndPush method
 *
 * @param {object} item Object of metadata for a given content item
 * @param {string} url The URL of the page to be rendered
 * @returns {array} Array of { property: content } key pairs
 */
export const getOgTags = ( item, url ) => {
  const {
    selectedLanguageUnit,
    title,
    type,
  } = item;

  const isVideo = type && type === 'video';

  /* TEMP vars for Document use case */
  const trimContent = item.content && item.content.rawText && `${item.content.rawText.substring( 0, 152 )}...`;
  const description = item.description || trimContent;

  const videoSrc = selectedLanguageUnit && selectedLanguageUnit.source && selectedLanguageUnit.source[0]
    ? selectedLanguageUnit.source[0]
    : {};
  const streamSrc = videoSrc.stream && videoSrc.stream.link ? videoSrc.stream.link : '';

  const meta = {
    'og:type': isVideo ? 'video.other' : 'article',
    'og:url': url,
    'twitter:card': isVideo ? 'player' : 'summary',
    'og:description': description,
    'og:title': title,
    'og:video:url': isVideo ? streamSrc : '',
    'twitter:player': isVideo ? streamSrc : '',
    'twitter:player:height': isVideo ? '196px' : '',
    'twitter:player:width': isVideo ? '350px' : '',
  };

  return filterEmpty( meta );
};
