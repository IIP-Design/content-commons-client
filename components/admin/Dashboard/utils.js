import { getProjectsType } from 'lib/graphql/util';
import {
  DELETE_GRAPHIC_PROJECT_MUTATION,
  TEAM_GRAPHIC_PROJECTS_QUERY,
  TEAM_GRAPHIC_PROJECTS_COUNT_QUERY,
  GRAPHIC_PROJECT_IMAGE_FILES_QUERY,
  UNPUBLISH_GRAPHIC_PROJECT_MUTATION,
  UPDATE_GRAPHIC_STATUS_MUTATION,
  GRAPHIC_PROJECTS_META_QUERY,
} from 'lib/graphql/queries/graphic';
import {
  DELETE_PACKAGE_MUTATION,
  PACKAGE_FILES_QUERY,
  TEAM_PACKAGES_COUNT_QUERY,
  TEAM_PACKAGES_QUERY,
  UNPUBLISH_PACKAGE_MUTATION,
  UPDATE_PACKAGE_STATUS_MUTATION,
  PACKAGES_META_QUERY,
} from 'lib/graphql/queries/package';
import {
  DELETE_VIDEO_PROJECT_MUTATION,
  TEAM_VIDEO_PROJECTS_COUNT_QUERY,
  TEAM_VIDEO_PROJECTS_QUERY,
  UNPUBLISH_VIDEO_PROJECT_MUTATION,
  UPDATE_VIDEO_STATUS_MUTATION,
  VIDEO_PROJECT_FILES_QUERY,
  VIDEO_PROJECTS_META_QUERY,
} from 'lib/graphql/queries/video';
import { TEAM_PLAYBOOKS_QUERY, TEAM_PLAYBOOKS_COUNT_QUERY } from 'lib/graphql/queries/playbook';

/**
 * Returns the project type based off of the content type available to a team
 *
 * @param {Object} team team data received from GraphQL
 */
export const getContentTypesFromTeam = team => {
  const contentTypes = team?.contentTypes ? team.contentTypes : '';

  const type = getProjectsType( contentTypes );

  return type || null;
};

/**
 * Returns the expected queries depending on what content types a team has access to
 *
 * @param {Object} team team data received from GraphQL
 * @returns {Object} list of relevant queries
 */
export const setQueries = team => {
  const queries = {};

  switch ( getContentTypesFromTeam( team ) ) {
    case 'graphicProjects':
      queries.content = TEAM_GRAPHIC_PROJECTS_QUERY;
      queries.count = TEAM_GRAPHIC_PROJECTS_COUNT_QUERY;
      queries.files = GRAPHIC_PROJECT_IMAGE_FILES_QUERY;
      queries.metaContent = GRAPHIC_PROJECTS_META_QUERY;
      queries.remove = DELETE_GRAPHIC_PROJECT_MUTATION;
      queries.status = UPDATE_GRAPHIC_STATUS_MUTATION;
      queries.unpublish = UNPUBLISH_GRAPHIC_PROJECT_MUTATION;

      return queries;
    case 'videoProjects':
      queries.content = TEAM_VIDEO_PROJECTS_QUERY;
      queries.count = TEAM_VIDEO_PROJECTS_COUNT_QUERY;
      queries.files = VIDEO_PROJECT_FILES_QUERY;
      queries.metaContent = VIDEO_PROJECTS_META_QUERY;
      queries.remove = DELETE_VIDEO_PROJECT_MUTATION;
      queries.status = UPDATE_VIDEO_STATUS_MUTATION;
      queries.unpublish = UNPUBLISH_VIDEO_PROJECT_MUTATION;

      return queries;
    case 'packages':
      queries.content = TEAM_PACKAGES_QUERY;
      queries.count = TEAM_PACKAGES_COUNT_QUERY;
      queries.files = PACKAGE_FILES_QUERY;
      queries.metaContent = PACKAGES_META_QUERY;
      queries.remove = DELETE_PACKAGE_MUTATION;
      queries.status = UPDATE_PACKAGE_STATUS_MUTATION;
      queries.unpublish = UNPUBLISH_PACKAGE_MUTATION;

      return queries;

    case 'playbooks':
      queries.content = TEAM_PLAYBOOKS_QUERY;
      queries.count = TEAM_PLAYBOOKS_COUNT_QUERY;
      // The following queries are placeholders and for the wrong content type
      // As such, these operations will not work properly
      queries.files = PACKAGE_FILES_QUERY;
      queries.metaContent = PACKAGES_META_QUERY;
      queries.remove = DELETE_PACKAGE_MUTATION;
      queries.status = UPDATE_PACKAGE_STATUS_MUTATION;
      queries.unpublish = UNPUBLISH_PACKAGE_MUTATION;

      return queries;
    default:
      return queries;
  }
};
