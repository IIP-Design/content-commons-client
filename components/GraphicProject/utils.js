import { getTransformedLangTaxArray, getCount } from 'lib/utils';

const structureLangObj = item => ( {
  language: {
    language_code: item.language.languageCode,
    text_direction: item.language.textDirection,
    locale: item.language.locale,
    display_name: item.language.displayName,
    native_name: item.language.nativeName,
  },
} );

export const normalizeGraphicProjectByAPI = ( { file, useGraphQl = false } ) => {
  const graphicObj = {
    id: file.id || '',
    type: 'graphic',
    site: file.site || '',
    title: file.title || '',
    projectType: file.projectType || '',
    published: file.published || '',
    modified: file.modified || '',
    owner: file.owner || '',
    desc: file.desc || '',
    descInternal: file.descInternal || '',
    copyright: file.copyright || '',
    images: file.images || [],
    supportFiles: file.supportFiles || [],
    categories: file.categories || '',
  };

  // Elastic API
  if ( !useGraphQl ) {
    const imagesWithFileNameProp = file.images.map( img => ( {
      ...img,
      filename: img.filename,
    } ) );

    const supportFilesWithFileNameProp = Array.isArray( file.supportFiles ) && file.supportFiles.map( supFile => {
      const filenameFromSrc = supFile.url.slice( supFile.url.lastIndexOf( '/' ) + 1 );

      return {
        ...supFile,
        filename: filenameFromSrc,
      };
    } );

    const esObj = {
      images: imagesWithFileNameProp || [],
      supportFiles: supportFilesWithFileNameProp || [],
    };

    return { ...graphicObj, ...esObj };
  }

  // GraphQL API
  if ( useGraphQl ) {
    const structuredImages = file.images.map( img => {
      const { dimensions } = img;

      return {
        ...img,
        ...structureLangObj( img ),
        url: img.signedUrl,
        width: dimensions.width,
        height: dimensions.height,
      };
    } );

    const structuredSupportFiles = Array.isArray( file.supportFiles ) && file.supportFiles.map( supFile => ( {
      ...supFile,
      ...structureLangObj( supFile ),
      srcUrl: supFile.signedUrl,
    } ) );

    const gqlObj = {
      alt: file.alt || '',
      projectType: file.type || '',
      published: file.publishedAt || '',
      modified: file.updatedAt || '',
      owner: file.team?.name || '',
      desc: file.descPublic || '',
      images: structuredImages || [],
      supportFiles: structuredSupportFiles || [],
      categories: getTransformedLangTaxArray( file.categories ) || [],
    };

    return { ...graphicObj, ...gqlObj };
  }

  return graphicObj;
};

// Use Twitter graphics as default for display otherwise whatever graphic image is available
export const filterGraphicImgs = images => {
  const containsTwitterImgs = images.some( img => img.social.includes( 'Twitter' ) );

  if ( containsTwitterImgs ) {
    return images.filter( img => img.social.includes( 'Twitter' ) );
  }

  return images;
};

export const getGraphicImgsBySocial = ( images, platform, useGraphQl = false ) => {
  if ( !getCount( images ) ) return [];

  const filteredImgs = images.reduce( ( allImgs, img ) => {
    let platformImgs = [];
    let condition = img.social.includes( platform );

    if ( useGraphQl ) {
      platformImgs = img.social.filter( s => s.name === platform );
      condition = getCount( platformImgs );
    }

    if ( condition ) {
      allImgs.push( img );
    }

    return allImgs;
  }, [] );

  return getCount( filteredImgs ) ? filteredImgs : images;
};
