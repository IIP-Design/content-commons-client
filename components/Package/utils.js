import { getTransformedLangTaxArray } from 'lib/utils';
import { packageDocumentsRequest } from 'lib/elastic/api';
import { getDataFromHits } from 'lib/elastic/parser';
import DosSeal from 'static/images/dos_seal.svg';

/*
 * Normalize Document type data structure from both APIs
 *
 */
export const normalizeDocumentItemByAPI = ( { file, useGraphQl = false } ) => {
  const documentObj = {
    type: 'document',
    id: file.id || '',
    published: file.published || '',
    modified: file.modified || '',
    author: file.author || '',
    owner: file.owner || '',
    site: file.site || '',
    title: file.title || '',
    excerpt: file.excerpt || '',
    content: file.content || '',
    logo: DosSeal || '',
    language: file.language,
    documentUrl: file.url || '',
    documentUse: file.use || '',
    tags: file.tags || [],
  };

  if ( useGraphQl ) {
    const graphFields = {
      created: file.createdAt || '',
      published: file.publishedAt || '',
      modified: file.updatedAt || '',
      author: file.author ? `${file.author.firstname} ${file.author.lastname}` : '',
      owner: file.team?.name || '',
      documentUrl: file.url || '',
      documentUse: file?.use?.name || '',
      excerpt: file?.excerpt || '',
      tags: getTransformedLangTaxArray( file.tags ) || [],
    };

    return { ...documentObj, ...graphFields };
  }

  return documentObj;
};

/*
 * Request documents for a package from Elasticsearch
 * @param docs array of objects w shape { id, item }
 */
export const getElasticPkgDocs = async ( documents, user ) => {
  if ( user && user.id !== 'public' ) {
    try {
      const docIds = documents.map( doc => doc.id );
      const responseHits = await packageDocumentsRequest( docIds, user );
      const docs = getDataFromHits( responseHits )?.map( hit => hit._source );

      return docs;
    } catch ( error ) {
      console.log( error );

      return error;
    }
  }
};
