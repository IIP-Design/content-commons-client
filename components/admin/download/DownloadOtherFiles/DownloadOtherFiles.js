import React from 'react';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';
import { Loader } from 'semantic-ui-react';
import DownloadItemContent from 'components/download/DownloadItem/DownloadItemContent';
import { capitalizeFirst, getFileExt, getS3Url } from 'lib/utils';
import ApolloError from 'components/errors/ApolloError';


const DownloadOtherFiles = props => {
  const { data, isPreview } = props;
  const { error, loading, project } = data;

  if ( loading ) {
    return (
      <div
        style={ {
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
          justifyContent: 'center',
          minHeight: '200px',
        } }
      >
        <Loader
          active
          inline="centered"
          style={ { marginBottom: '1em' } }
          content="Loading other file(s)..."
        />
      </div>
    );
  }

  if ( error ) return <ApolloError error={ error } />;
  if ( !project || !Object.keys( project ).length ) return null;

  const { files, thumbnails } = project;
  const allFiles = [...thumbnails, ...files];

  const getInstructions = ( { language, filename, filetype } ) => {
    const isImage = filetype.startsWith( 'image/' );
    const extension = getFileExt( filename, false );
    const capitalizedExt = capitalizeFirst( extension );

    if ( isImage ) {
      return `Download ${language} Thumbnail`;
    }

    return `Download "${filename}" (${capitalizedExt})`;
  };

  const renderFormItem = file => {
    const {
      id, filename, filetype, url,
      language: { displayName },
    } = file;

    return (
      <DownloadItemContent
        key={ `fs_${id}` }
        srcUrl={ getS3Url( url ) }
        hoverText={ getInstructions( { language: displayName, filename, filetype } ) }
        isAdminPreview={ isPreview }
      >
        <div className="item-content">
          <p className="item-content__title">
            <strong>
              { getInstructions( { language: displayName, filename, filetype } ) }
            </strong>
          </p>
        </div>
      </DownloadItemContent>
    );
  };

  const renderFormItems = () => {
    const otherFiles = allFiles
      .filter( file => file && file.url )
      .map( file => renderFormItem( file ) );

    return otherFiles.length ? otherFiles : 'There are no other files available for download at this time';
  };

  return files && renderFormItems( files );
};

DownloadOtherFiles.propTypes = {
  data: PropTypes.object,
  isPreview: PropTypes.bool,
};

const VIDEO_PROJECT_PREVIEW_OTHER_FILES_QUERY = gql`
  query VideoProjectPreviewOtherFiles($id: ID!) {
    project: videoProject(id: $id) {
      id
      files: supportFiles(
        where: {
          AND: [
            { filetype_not: "application/x-subrip" },
            { filename_not_ends_with: "vtt" }
          ]
        },
        orderBy: filename_ASC
      ) {
        id
        filename
        filetype
        url
        language {
          id
          displayName
        }
      }
      thumbnails(
        where: {
          OR: [
            { filetype: "image/jpeg" },
            { filetype: "image/png" },
            { filename_ends_with: "jpg" },
            { filename_ends_with: "png" }
          ],
          AND: {
            use: { name: "Thumbnail/Cover Image" }
          }
        },
        orderBy: filename_ASC
      ) {
        id
        filename
        filetype
        url
        language {
          id
          displayName
        }
      }
    }
  }
`;

export default graphql( VIDEO_PROJECT_PREVIEW_OTHER_FILES_QUERY, {
  options: props => ( {
    variables: {
      id: props.id,
    },
  } ),
} )( DownloadOtherFiles );

export { VIDEO_PROJECT_PREVIEW_OTHER_FILES_QUERY };
