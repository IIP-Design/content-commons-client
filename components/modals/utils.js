import GraphicProject from 'components/GraphicProject/GraphicProject';
import Post from 'components/Post/Post';
import Video from 'components/Video/Video';
import Document from 'components/Document/Document';
/**
 * Check content type of an item and passes in the correct component to the modal
 * @param {Object} item data from the selected item
 */
export const getModalContent = item => {
  const noContent = <div>No content currently available</div>;

  if ( item ) {
    switch ( item.type ) {
      case 'graphic':
        return <GraphicProject item={ item } displayAsModal />;

      case 'video':
        return <Video item={ item } />;

      case 'post':
        return <Post item={ item } />;

      case 'document':
        return <Document item={ item } displayAsModal />;

      default:
        return noContent;
    }
  }

  return noContent;
};
