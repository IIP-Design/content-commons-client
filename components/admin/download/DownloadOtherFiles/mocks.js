import { VIDEO_PROJECT_PREVIEW_OTHER_FILES_QUERY } from './DownloadOtherFiles';

export const props = {
  id: '123',
  instructions: 'Download Other File(s)',
  isPreview: false,
};

export const mocks = [
  {
    request: {
      query: VIDEO_PROJECT_PREVIEW_OTHER_FILES_QUERY,
      variables: { id: props.id },
    },
    result: {
      data: {
        project: {
          id: props.id,
          files: [
            {
              id: 'pdf89',
              filename: 'file-1.pdf',
              filetype: 'application/pdf',
              url: `2019/06/${props.id}/file-1.pdf`,
              language: {
                id: 'en23',
                displayName: 'English',
              },
            },
            {
              id: 'au56',
              filename: 'audio-1.mp3',
              filetype: 'audio/x-mpeg-3',
              url: `2019/06/${props.id}/audio-1.mp3`,
              language: {
                id: 'en23',
                displayName: 'English',
              },
            },
          ],
          thumbnails: [
            {
              id: 'th765',
              filename: 'thumbnail-1.jpg',
              filetype: 'image/jpeg',
              url: `2019/06/${props.id}/thumbnail-1.jpg`,
              language: {
                id: 'en23',
                displayName: 'English',
              },
            },
            {
              id: 'th865',
              filename: 'thumbnail-2.jpg',
              filetype: 'image/jpeg',
              url: `2019/06/${props.id}/thumbnail-2.jpg`,
              language: {
                id: 'en23',
                displayName: 'English',
              },
            },
          ],
        },
      },
    },
  },
];

export const errorMocks = [
  {
    ...mocks[0],
    result: {
      errors: [{ message: 'There was an error.' }],
    },
  },
];

export const nullProjectMocks = [
  {
    ...mocks[0],
    result: { data: { project: null } },
  },
];

export const emptyProjectMocks = [
  {
    ...mocks[0],
    result: { data: { project: {} } },
  },
];

export const noFilesMocks = [
  {
    ...mocks[0],
    result: {
      data: {
        project: {
          ...mocks[0].result.data.project,
          files: [],
          thumbnails: [],
        },
      },
    },
  },
];
