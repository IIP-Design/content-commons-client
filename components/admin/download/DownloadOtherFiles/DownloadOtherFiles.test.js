import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from '@apollo/react-testing';
import { Loader } from 'semantic-ui-react';

import DownloadOtherFiles from './DownloadOtherFiles';

import { suppressActWarning } from 'lib/utils';
import {
  emptyProjectMocks,
  errorMocks,
  mocks,
  noFilesMocks,
  nullProjectMocks,
  props,
} from './mocks';

jest.mock( 'lib/hooks/useSignedUrl', () => jest.fn( () => ( { signedUrl: 'https://example.jpg' } ) ) );

jest.mock( 'lib/utils', () => ( {
  ...jest.requireActual( '../../../../lib/utils' ),
  getS3Url: jest.fn( assetPath => `https://s3-url.com/${assetPath}` ),
  getApolloErrors: error => {
    let errs = [];
    const { graphQLErrors, networkError, otherError } = error;

    if ( graphQLErrors ) {
      errs = graphQLErrors.map( error => error.message );
    }
    if ( networkError ) {
      errs.push( networkError );
    }

    if ( otherError ) {
      errs.push( otherError );
    }

    return errs;
  },
} ) );

jest.mock( 'static/icons/icon_download.svg', () => 'downloadIconSVG' );

const Component = (
  <MockedProvider mocks={ mocks } addTypename={ false }>
    <DownloadOtherFiles { ...props } />
  </MockedProvider>
);

describe( '<DownloadOtherFiles />', () => {
  const consoleError = console.error;

  beforeAll( () => suppressActWarning( consoleError ) );
  afterAll( () => {
    console.error = consoleError;
  } );

  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const downloadOther = wrapper.find( 'DownloadOtherFiles' );
    const loader = (
      <Loader
        active
        inline="centered"
        style={ { marginBottom: '1em' } }
        content="Loading other file(s)..."
      />
    );

    expect( downloadOther.exists() ).toEqual( true );
    expect( downloadOther.contains( loader ) ).toEqual( true );
    expect( toJSON( downloadOther ) ).toMatchSnapshot();
  } );

  it( 'renders error message if error is thrown', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ errorMocks } addTypename={ false }>
        <DownloadOtherFiles { ...props } />
      </MockedProvider>,
    );

    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const downloadOther = wrapper.find( 'DownloadOtherFiles' );
    const errorComponent = downloadOther.find( 'ApolloError' );

    expect( errorComponent.exists() ).toEqual( true );
    expect( errorComponent.contains( 'There was an error.' ) )
      .toEqual( true );
  } );

  it( 'renders final state without crashing', async () => {
    const wrapper = mount( Component );

    await wait( 0 );
    wrapper.update();

    const downloadOther = wrapper.find( 'DownloadOtherFiles' );

    expect( toJSON( downloadOther ) ).toMatchSnapshot();
  } );

  it( 'renders null if project is null', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ nullProjectMocks } addTypename={ false }>
        <DownloadOtherFiles { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const downloadOther = wrapper.find( 'DownloadOtherFiles' );

    expect( downloadOther.html() ).toEqual( null );
  } );

  it( 'renders null if project is {}', async () => {
    // ignore console.warn about missing field `id` and `files`
    const consoleWarn = console.warn;

    console.warn = jest.fn();

    const wrapper = mount(
      <MockedProvider mocks={ emptyProjectMocks } addTypename={ false }>
        <DownloadOtherFiles { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const downloadOther = wrapper.find( 'DownloadOtherFiles' );

    expect( downloadOther.html() ).toEqual( null );
    console.warn = consoleWarn;
  } );

  it( 'renders a "no other files available message" if there are no other files', async () => {
    const wrapper = mount(
      <MockedProvider mocks={ noFilesMocks } addTypename={ false }>
        <DownloadOtherFiles { ...props } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const downloadOther = wrapper.find( 'DownloadOtherFiles' );
    const items = downloadOther.find( 'Item' );
    const { files, thumbnails } = noFilesMocks[0].result.data.project;
    const allFiles = [...thumbnails, ...files];
    const msg = 'There are no other files available for download at this time';

    expect( downloadOther.exists() ).toEqual( true );
    expect( items.length ).toEqual( allFiles.length );
    expect( downloadOther.contains( msg ) ).toEqual( true );
  } );

  it( 'renders <a> tags with the correct href and download attribute values', async () => {
    const wrapper = mount( Component );

    await wait( 0 );
    wrapper.update();

    const items = wrapper.find( 'DownloadItemContent' );
    const { files, thumbnails } = mocks[0].result.data.project;
    const allFiles = [...thumbnails, ...files];
    const s3Bucket = 'https://s3-url.com';

    expect( items.length ).toEqual( allFiles.length );
    items.forEach( ( item, i ) => {
      const { url: assetPath } = allFiles[i];

      expect( item.prop( 'srcUrl' ) ).toEqual( `${s3Bucket}/${assetPath}` );
      expect( item.find( '.download-item' ).prop( 'href' ) ).toEqual( 'https://example.jpg' );
      expect( item.find( '.download-item' ).prop( 'download' ) ).toEqual( true );
    } );
  } );

  it( 'renders preview text if isPreview is true', async () => {
    const newProps = { ...props, isPreview: true };
    const wrapper = mount(
      <MockedProvider mocks={ mocks } addTypename={ false }>
        <DownloadOtherFiles { ...newProps } />
      </MockedProvider>,
    );

    await wait( 0 );
    wrapper.update();

    const downloadOtherFiles = wrapper.find( 'DownloadOtherFiles' );
    const previews = downloadOtherFiles.find( '.preview-text' );
    const { files, thumbnails } = mocks[0].result.data.project;
    const allFiles = [...thumbnails, ...files];

    expect( previews.length ).toEqual( allFiles.length );
    previews.forEach( preview => {
      expect( preview.exists() )
        .toEqual( downloadOtherFiles.prop( 'isPreview' ) );
      expect( preview.name() ).toEqual( 'span' );
      expect( preview.text() )
        .toEqual( 'The link will be active after publishing.' );
    } );
  } );

  it( 'does not render preview text if !isPreview', async () => {
    const wrapper = mount( Component );

    await wait( 0 );
    wrapper.update();

    const downloadOtherFiles = wrapper.find( 'DownloadOtherFiles' );
    const previews = downloadOtherFiles.find( '.preview-text' );

    expect( previews.exists() )
      .toEqual( downloadOtherFiles.prop( 'isPreview' ) );
    expect( previews.length ).toEqual( 0 );
  } );
} );
