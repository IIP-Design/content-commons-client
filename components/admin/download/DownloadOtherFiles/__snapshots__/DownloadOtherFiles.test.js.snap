// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DownloadOtherFiles /> renders final state without crashing 1`] = `
<DownloadOtherFiles
  data={
    Object {
      "called": true,
      "error": undefined,
      "fetchMore": [Function],
      "loading": false,
      "networkStatus": 7,
      "project": Object {
        "files": Array [
          Object {
            "filename": "file-1.pdf",
            "filetype": "application/pdf",
            "id": "pdf89",
            "language": Object {
              "displayName": "English",
              "id": "en23",
            },
            "url": "2019/06/123/file-1.pdf",
          },
          Object {
            "filename": "audio-1.mp3",
            "filetype": "audio/x-mpeg-3",
            "id": "au56",
            "language": Object {
              "displayName": "English",
              "id": "en23",
            },
            "url": "2019/06/123/audio-1.mp3",
          },
        ],
        "id": "123",
        "thumbnails": Array [
          Object {
            "filename": "thumbnail-1.jpg",
            "filetype": "image/jpeg",
            "id": "th765",
            "language": Object {
              "displayName": "English",
              "id": "en23",
            },
            "url": "2019/06/123/thumbnail-1.jpg",
          },
          Object {
            "filename": "thumbnail-2.jpg",
            "filetype": "image/jpeg",
            "id": "th865",
            "language": Object {
              "displayName": "English",
              "id": "en23",
            },
            "url": "2019/06/123/thumbnail-2.jpg",
          },
        ],
      },
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "123",
      },
    }
  }
  id="123"
  instructions="Download Other File(s)"
  isPreview={false}
>
  <DownloadItemContent
    hoverText="Download English Thumbnail"
    isAdminPreview={false}
    key="fs_th765"
    srcUrl="https://s3-url.com/2019/06/123/thumbnail-1.jpg"
  >
    <a
      className="download-item"
      download={true}
      href="https://example.jpg"
      key="https://example.jpg"
      rel="noopener noreferrer"
    >
      <div
        className="item-icon"
      >
        <img
          alt="Download English Thumbnail"
          src="downloadIconSVG"
        />
      </div>
      <div
        className="item-content"
      >
        <p
          className="item-content__title"
        >
          <strong>
            Download English Thumbnail
          </strong>
        </p>
      </div>
      <span
        className="item-hover"
      >
        Download English Thumbnail
      </span>
    </a>
  </DownloadItemContent>
  <DownloadItemContent
    hoverText="Download English Thumbnail"
    isAdminPreview={false}
    key="fs_th865"
    srcUrl="https://s3-url.com/2019/06/123/thumbnail-2.jpg"
  >
    <a
      className="download-item"
      download={true}
      href="https://example.jpg"
      key="https://example.jpg"
      rel="noopener noreferrer"
    >
      <div
        className="item-icon"
      >
        <img
          alt="Download English Thumbnail"
          src="downloadIconSVG"
        />
      </div>
      <div
        className="item-content"
      >
        <p
          className="item-content__title"
        >
          <strong>
            Download English Thumbnail
          </strong>
        </p>
      </div>
      <span
        className="item-hover"
      >
        Download English Thumbnail
      </span>
    </a>
  </DownloadItemContent>
  <DownloadItemContent
    hoverText="Download \\"file-1.pdf\\" (Pdf)"
    isAdminPreview={false}
    key="fs_pdf89"
    srcUrl="https://s3-url.com/2019/06/123/file-1.pdf"
  >
    <a
      className="download-item"
      download={true}
      href="https://example.jpg"
      key="https://example.jpg"
      rel="noopener noreferrer"
    >
      <div
        className="item-icon"
      >
        <img
          alt="Download \\"file-1.pdf\\" (Pdf)"
          src="downloadIconSVG"
        />
      </div>
      <div
        className="item-content"
      >
        <p
          className="item-content__title"
        >
          <strong>
            Download "file-1.pdf" (Pdf)
          </strong>
        </p>
      </div>
      <span
        className="item-hover"
      >
        Download "file-1.pdf" (Pdf)
      </span>
    </a>
  </DownloadItemContent>
  <DownloadItemContent
    hoverText="Download \\"audio-1.mp3\\" (Mp3)"
    isAdminPreview={false}
    key="fs_au56"
    srcUrl="https://s3-url.com/2019/06/123/audio-1.mp3"
  >
    <a
      className="download-item"
      download={true}
      href="https://example.jpg"
      key="https://example.jpg"
      rel="noopener noreferrer"
    >
      <div
        className="item-icon"
      >
        <img
          alt="Download \\"audio-1.mp3\\" (Mp3)"
          src="downloadIconSVG"
        />
      </div>
      <div
        className="item-content"
      >
        <p
          className="item-content__title"
        >
          <strong>
            Download "audio-1.mp3" (Mp3)
          </strong>
        </p>
      </div>
      <span
        className="item-hover"
      >
        Download "audio-1.mp3" (Mp3)
      </span>
    </a>
  </DownloadItemContent>
</DownloadOtherFiles>
`;

exports[`<DownloadOtherFiles /> renders initial loading state without crashing 1`] = `
<DownloadOtherFiles
  data={
    Object {
      "called": true,
      "error": undefined,
      "fetchMore": [Function],
      "loading": true,
      "networkStatus": 1,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "123",
      },
    }
  }
  id="123"
  instructions="Download Other File(s)"
  isPreview={false}
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "minHeight": "200px",
      }
    }
  >
    <Loader
      active={true}
      content="Loading other file(s)..."
      inline="centered"
      style={
        Object {
          "marginBottom": "1em",
        }
      }
    >
      <div
        className="ui active text centered inline loader"
        style={
          Object {
            "marginBottom": "1em",
          }
        }
      >
        Loading other file(s)...
      </div>
    </Loader>
  </div>
</DownloadOtherFiles>
`;
